### PR TITLE
offsets cleanup

### PR DESCRIPTION
**Pull request by jbrockmendel**
This is mostly a small cleanup of offsets, want to keep this fairly trivial diff from obscuring more important upcoming changes.

The two non-trivial things here are 1) implement `_Tick` class in offsets.pyx with the most basic of its methods and 2) optimization of `Tick.__hash__`.  You'd be surprised how big a difference this particular optimization makes:

Before:
```
In [2]: day = pd.offsets.Day(2)
In [3]: %timeit hash(day)
100000 loops, best of 3: 11.7 µs per loop
```

After
```
In [3]: day = pd.offsets.Day(2)
In [5]: %timeit hash(day)
The slowest run took 8.61 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 1.86 µs per loop
```

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -375,3 +376,21 @@ class BaseOffset(_BaseOffset):
             # i.e. isinstance(other, (ABCDatetimeIndex, ABCSeries))
             return other - self
         return -self + other
+
+# ---------------------------------------------------------------------
+# Ticks
+
+class _Tick(object):
+    _inc = Timedelta(microseconds=1000)
+    _prefix = 'undefined'
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: i would make this None as it would hard fail if not defined (or you can make it NotImemenrsError)

**Review by jreback — CHANGES_REQUESTED** [review]
add your examples as an asv (and or run them and show the existing ones as a diff); the freq ones

**Comment by jreback** [comment]
also add a note in perf for 0.22; as you make more changes u can add those issue numbers in (PR numbers)

**Comment by jbrockmendel** [comment]
There's a lot of upcoming work in offsets.  Luckily it isn't all cut/paste, so it won't be quite the deluge of PRs.  Going to put together a thorough set of asvs; will remove the optimization and make this pure-cleanup for now.

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -1227,12 +1227,7 @@ def onOffset(self, dt):
         if self.normalize and not _is_normalized(dt):
             return False
         first_weekday, _ = tslib.monthrange(dt.year, dt.month)
-        if first_weekday == 5:
-            return dt.day == 3
```
Comment (jreback) [inline pandas/tseries/offsets.py]: is this the same logic?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
Yes

**Comment by jreback** [comment]
linting issue & rebase

**Comment by jreback** [comment]
would like to have some benchmark asv's here as well.

**Pull request closed without merging**